### PR TITLE
[TOAZ-315] Add Leo app health metrics reporter

### PR DESCRIPTION
**Pull request by rtitle**
https://broadworkbench.atlassian.net/browse/TOAZ-315

Add Leo metrics for app health. The rough design is:

* New background process `LeoMetricsMonitor` in back Leo (this actually replaces the old defunct `ClusterToolMonitor` class)
* Periodically scans the DB for all active runtimes and apps (on both clouds)
* Health checks all running runtimes/apps, and send metrics to prometheus/grafana about their stauts/health

See inline for more details, but that's the rough idea.

For a demo, I actually enabled it on dev for a short time and created this Grafana dashboard: https://grafana.dsp-devops.broadinstitute.org/d/mmgNHw04z/leonardo-apps?from=1675914426775&to=1675957628933&orgId=1

Still getting used to Grafana but it displays counts of runtimes/apps broken down in various ways. I believe it is accurate comparing with dev DB data. Looking forward to seeing the metrics for prod!

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAO.scala [hunk 1]
```diff
@@ -32,22 +31,30 @@ class HttpCromwellDAO[F[_]](httpClient: Client[F])(implicit
...
 
   override def getStatus(baseUri: Uri, authHeader: Authorization)(implicit
     ev: Ask[F, AppContext]
   ): F[Boolean] =
     for {
       _ <- metrics.incrementCounter("cromwell/status")
-      cromwellStatusUri = baseUri / "cromwell" / "api" / "engine" / "v1" / "status"
-      res <- httpClient.expectOr[CromwellStatusCheckResponse](
+      cromwellStatusUri = baseUri / "cromwell" / "engine" / "v1" / "status"
+      // Note: cromwell-as-an-app /status returns '{}' in the response body for some reason.
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAO.scala:41]: Do you think it is worth creating a follow up ticket for? It might fall into limbo otherwise and we will forget to revisit

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala [hunk 2]
```diff
@@ -511,9 +511,12 @@ object MemorySize {
  */
 final case class RuntimeResourceConstraints(memoryLimit: MemorySize)
 
-final case class RunningRuntime(cloudContext: CloudContext,
```
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala:514]: `RunningRuntime` was previously used by `ClusterToolMonitor` (which I deleted) -- renamed to `RuntimeMetrics` and added some things.

File: http/src/main/resources/reference.conf [hunk 3]
```diff
@@ -975,3 +975,8 @@ app-service {
 drs {
   url = "https://drshub.dsde-dev.broadinstitute.org/api/v4/drs/resolve"
 }
+
+metrics {
+  check-interval = 5 minutes
+  include-azure-cloud-context = true
```
Comment (rtitle) [inline http/src/main/resources/reference.conf:985]: Normally it's not good to add metrics for things with high cardinality. We wouldn't want to include a tag for `workspaceId`, for example. However MRGs are pretty low cardinality, and per-MRG metrics would be extremely useful for public preview launch. So I'm including it as a tag for now, but making it configurable.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 4]
```diff
@@ -242,10 +242,17 @@ object clusterQuery extends TableQuery(new ClusterTable(_)) {
...
+  // left join label l on c.id = l.clusterId
+  val clusterMetricsQuery: Query[(ClusterTable, ClusterImageTable, Rep[Option[LabelTable]]),
+                                 (ClusterRecord, ClusterImageRecord, Option[LabelRecord]),
+                                 Seq
+  ] =
     for {
-      (cluster, image) <- clusterQuery join clusterImageQuery on (_.id === _.clusterId)
-    } yield (cluster, image)
+      ((cluster, image), label) <-
+        clusterQuery join clusterImageQuery on (_.id === _.clusterId) joinLeft labelQuery on { case ((c, _), lbl) =>
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:252]: Hopefully joining runtimes to labels in the metrics query is okay.. we use the labels for distinguishing Terra vs AoU runtimes, which is useful info. It seemed to perform okay on dev.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterToolMonitor.scala [hunk 5]
```diff
@@ -1,88 +0,0 @@
...
-import com.typesafe.scalalogging.LazyLogging
-import org.broadinstitute.dsde.workbench.leonardo.config.ClusterToolConfig
-import org.broadinstitute.dsde.workbench.leonardo.dao.ToolDAO
-import org.broadinstitute.dsde.workbench.leonardo.db.{clusterQuery, DbReference}
-import org.broadinstitute.dsde.workbench.leonardo.monitor.ClusterToolMonitor._
-import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
-
-import scala.concurrent.ExecutionContext
-
-object ClusterToolMonitor {
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterToolMonitor.scala:17]: As mentioned in the description, this class had a similar intent to `LeoMetricsMonitor` but (a) it was checking GCP/Jupyter only; and (b) I don't think it was actually running. :) 

I also was surprised to see this was an akka `Actor`… I thought we had converted them all to `fs2`.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.http4s.headers.Authorization
+import org.http4s.{AuthScheme, Credentials, Uri}
+import org.typelevel.log4cats.StructuredLogger
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+
+/** Collects metrics about active Leo runtimes and apps. */
+class LeoMetricsMonitor[F[_]](config: LeoMetricsMonitorConfig,
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:27]: All the metrics collection logic is in this class

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+                      s"Unexpected app service encountered during health checks: ${s.value}"
+                    ) >> F.pure(false)
+                }
+              } yield isUp
+          }
+          // In addition to collecting aggregate metrics, log a warning for any app that is down.
+          _ <-
+            if (isUp) F.unit
+            else
+              logger.error(ctx.loggingCtx)(
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:171]: Right now this is just writing a log line -- my idea was we could hook this up to errorReporting or sentry in the future if we want to alert on apps that are down.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  private def getAzureCloudContext(cloudContext: CloudContext): Option[AzureCloudContext] =
+    (config.includeAzureCloudContext, cloudContext) match {
+      case (true, CloudContext.Azure(cc)) => Some(cc)
+      case _                              => None
+    }
+}
+
+case class LeoMetricsMonitorConfig(checkInterval: FiniteDuration, includeAzureCloudContext: Boolean)
+
+sealed trait LeoMetric {
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:275]: Little ADT for Leo metrics. A metric has a name; a type (we are always using `gauge` in this class but there are also `counter` and `timer); tags; and a value.

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitorSpec.scala [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+                            RuntimeImageType.stringToRuntimeImageType(s),
+                            RuntimeUI.Terra,
+                            None,
+                            s == "RStudio"
+        )
+      ) shouldBe Some(0)
+    }
+  }
+
+  // Data generators
```
Comment (rtitle) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitorSpec.scala:253]: Was half thinking this test would be a good candidate for Scala [property based testing](https://scalacheck.org/) but I just hardcoded some seed data.

File: http/src/main/resources/reference.conf (hunk 3 above)
Comment (LizBaldo) [reply http/src/main/resources/reference.conf:985]: That makes sense, given this context it would be nice to move that comment from github into this file for the next dev who will update this :)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/CbasUiDAO.scala [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package org.broadinstitute.dsde.workbench.leonardo.dao
+
+import cats.mtl.Ask
+import org.broadinstitute.dsde.workbench.leonardo.AppContext
+import org.http4s.Uri
+import org.http4s.headers.Authorization
+
+trait CbasUiDAO[F[_]] {
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/CbasUiDAO.scala:8]: dao stands for `decentralized autonomous organization` in this context correct?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 4 above)
Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:252]: It makes sense to me that returning the label would be useful for filtering metrics downs the road (I have some questions about this label table and why we have it but that is way outside of the scope of this PR 😅)
Leo's tables are not crazy big so adding a little left join should not impact performance 🤞

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 11]
```diff
@@ -414,9 +421,9 @@ object clusterQuery extends TableQuery(new ClusterTable(_)) {
       scopes <- scopeQuery.getAllForCluster(runtimeId)
     } yield ExtraInfoForCreateRuntime(images.toSet, extention, scopes)
 
-  def listRunningOnly(implicit ec: ExecutionContext): DBIO[Seq[RunningRuntime]] =
-    clusterJoinClusterImageQuery.filter(_._1.status === (RuntimeStatus.Running: RuntimeStatus)).result map {
-      unmarshalRunningCluster
+  def listActiveForMetrics(implicit ec: ExecutionContext): DBIO[Seq[RuntimeMetrics]] =
+    clusterMetricsQuery.filter(_._1.status inSetBind RuntimeStatus.activeStatuses).result map {
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:425]: nifty! I did not know about inSetBind and just googled it, TIL :)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 12]
```diff
@@ -414,9 +421,9 @@ object clusterQuery extends TableQuery(new ClusterTable(_)) {
...
+      unmarshalRuntimeForMetrics
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:426]: I keep reminding myself that marshalling is a sort of serialization 😬

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala [hunk 13]
```diff
@@ -257,14 +257,30 @@ object Boot extends IOApp {
               appDependencies.asyncTasksQueue
             )
 
+          implicit val clusterToolToToolDao =
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala:263]: I might be a bit confused by the naming convention, but it looks like runtime-related metrics are bundled into the `clusterToolToToolDao` objec, while the metrics Monitor only contains app-related information?

I would expect something named `metricsMonitor` to report on every metric, not just the app-related ones, am I missing something?

Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala:263]: I just had to keep scrolling until the actual definition of the leo metrics monitor, I think I get it, but a little comment explaining the difference might be nice

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+  val process: Stream[F, Unit] =
+    (Stream.sleep[F](config.checkInterval) ++ Stream.eval(
+      retrieveMetrics
+        .handleErrorWith(e => logger.error(e)("Unexpected error occurred during metric monitoring"))
+    )).repeat
+
+  private[monitor] def retrieveMetrics: F[Unit] =
+    for {
+      now <- F.realTimeInstant
+      traceId = TraceId(s"AppMetricsMonitor_${now.toEpochMilli}")
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:56]: Nit, but why wouldn't we get a traceId back for runtimes as well?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+      c <- allClusters
+      n <- c.nodepools
+      a <- n.apps
+    } yield Map(
+      AppStatusMetric(c.cloudContext.cloudProvider,
+                      a.appType,
+                      a.status,
+                      getRuntimeUI(a.labels),
+                      getAzureCloudContext(c.cloudContext)
+      ) -> 1
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:99]: Noob question, but why `-> 1`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+    )
+
+    allApps.combineAll
+  }
+
+  /** Transforms DB runtimes to RuntimeStatusMetric and computes counts. */
+  private[monitor] def countRuntimesByDbStatus(allRuntimes: List[RuntimeMetrics]): Map[RuntimeStatusMetric, Int] = {
+    val allContainers = for {
+      r <- allRuntimes
+      c <- r.containers if c != RuntimeContainerServiceType.WelderService
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:113]: Why are we excluding welder here?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+      a <- n.apps if a.status == AppStatus.Running
+      s <- a.appResources.services
+    } yield (c.cloudContext, c.asyncFields.get.loadBalancerIp, a, s.config.name)
+
+    allServices
+      .parTraverseN(parallelism) { case (cloudContext, baseUri, app, serviceName) =>
+        for {
+          ctx <- ev.ask
+          // For GCP just test if the app is available through the Leo proxy.
+          // For Azure impersonate the user and call the app's status endpoint via Azure Relay.
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:143]: Oh whoa, naively this looks a lot more complex / expensie to do

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+                token <- F.fromOption(
+                  tokenOpt,
+                  AppCreationException(s"Pet not found for user ${app.auditInfo.creator}", Some(ctx.traceId))
+                )
+                authHeader = Authorization(Credentials.Token(AuthScheme.Bearer, token))
+                relayPath = Uri.unsafeFromString(baseUri.asString) / app.appName.value
+                isUp <- serviceName match {
+                  case ServiceName("wds")      => wdsDAO.getStatus(relayPath, authHeader).handleError(_ => false)
+                  case ServiceName("cbas")     => cbasDAO.getStatus(relayPath, authHeader).handleError(_ => false)
+                  case ServiceName("cbas-ui")  => cbasUiDAO.getStatus(relayPath, authHeader).handleError(_ => false)
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:158]: Noob question again, but why do we only separate the ui portion for cbas?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 7 above)
Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:171]: That would be pretty awesome!
Counting apps that are down could be useful to display on grafana as well though

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+                              samDAO: SamDAO[F]
+)(implicit
+  F: Async[F],
+  dbRef: DbReference[F],
+  metrics: OpenTelemetryMetrics[F],
+  logger: StructuredLogger[F],
+  clusterToolToToolDao: RuntimeContainerServiceType => ToolDAO[F, RuntimeContainerServiceType],
+  ec: ExecutionContext
+) {
+  private val parallelism = 25
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:42]: Where does the 25 come from?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitorSpec.scala (hunk 9 above)
Comment (LizBaldo) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitorSpec.scala:253]: Okay, I am not super familiar with this yet, but this feels like something that could potentially be reused elsewhere when testing and could therefore not be private?

**Review by LizBaldo — APPROVED** [review]
I left a lot of questions (sorry), mostly for my own understanding. The refactoring looks amazing to me, I think my main comment is around whether we should keep a counter for resources that are down in addition to logging to stay consistent with what the `ClusterToolMonitor` was trying to do.

File: http/src/main/resources/reference.conf (hunk 3 above)
Comment (rtitle) [reply http/src/main/resources/reference.conf:985]: Yeah sounds good! Will do.

Also crossed my mind to make the `LeoMetricsMonitor` as a whole disable-able -- just in case of disaster. :)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/CbasUiDAO.scala (hunk 10 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/CbasUiDAO.scala:8]: Data access object

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/CbasUiDAO.scala:8]: https://en.wikipedia.org/wiki/Data_access_object -- just a convention, though typically in the past I've seen DAO associated with database access, not so much web services

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAO.scala (hunk 1 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAO.scala:41]: Yes probably - I meant to ping batch team but slipped my mind. Can make a ticket.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 4 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:252]: Makes sense. Yeah I think the tables are indexed correctly so it should be okay.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala (hunk 13 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala:263]: Yeah the naming is maybe not ideal. :( `LeoMetricsMonitor` deals with both runtimes and apps and reports both kinds of metrics. `clusterToolToToolDao` and `ToolDAO` deals just with runtimes. The `WdsDAO`, `CromwellDAO`, etc are apps. Wonder if this could be made clearer with comments or something..

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 19 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:42]: A guess! Seemed like a reasonable parallelism limit. :-)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 14 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:56]: This is generating a TraceId to be included in logs by the `LeoMetricsMonitor`. This is sort of the entry point of the flow -- the trace id goes on to be passed to `retrieveAppMetrics` and `retrieveRuntimeMetrics` implicitly.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 15 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:99]: Not a noob question! Maybe worth a comment to explain.

The for-comp is building up a `List[Map[metric -> 1]]` structure. Then the `combineAll` is "folding" that structure into a `Map[metric -> n]`. If you want to get all FP about it it's using a [Monoid](https://typelevel.org/cats/typeclasses/monoid.html) instance for List, Map, Int behind the scenes. :-D

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 16 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:113]: Ah, because I wanted to have 1 `RuntimeStatusMetric` per runtime. But each runtime has a set of image types: `([jupyter or rstudio], welder)`. For this metric I just cared about reporting whether a runtime is jupyter or rstudio.

…I guess this relies on some non-intuitive assumptions of Leo internals.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 17 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:143]: It is! Mostly because of the Azure relay setup, as opposed to the Leo proxy. Perhaps we could have some utilities in code to simplify this.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 18 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:158]: Because only cbas-app has a separate UI. The others are just Java services.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 7 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:171]: Yup, I think both are complementary

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala (hunk 16 above)
Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala:113]: Ahhh gotcha, that makes a lot of sense, maybe a small comment would help future devs? There is so much implicit context in the inner workings of leo that I have yet to grasp 😅

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAOSpec.scala [hunk 20]
```diff
@@ -42,7 +42,7 @@ class HttpCromwellDAOSpec extends AnyFlatSpec with Matchers with LeonardoTestSui
       """.stripMargin
 
     val okCrom = Client.fromHttpApp[IO](
-      HttpApp(_ => IO.fromEither(parse(response)).flatMap(r => IO(Response(status = Status.Ok).withEntity(r))))
+      HttpApp(_ => IO.fromEither(parse(response)).flatMap(r => IO(Response(status = Status.BadGateway).withEntity(r))))
     )
```
Comment (lucymcnatt) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAOSpec.scala:46]: Why was this changed from `Ok` to `BadGateway` here?

Comment (rtitle) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAOSpec.scala:46]: Ah, see the TODO in `HttpCromwellDAO`. This test is trying to simulate an unhealthy cromwell. But now we treat any OK status as healthy. So updating the test to return 502 (which is actually the status code we get when the cromwell pod is down).

**Comment by rtitle** [comment]
jenkins retest, RStudio swatomation test failed

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAOSpec.scala (hunk 20 above)
Comment (lucymcnatt) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpCromwellDAOSpec.scala:46]: ohh ok I didn't connect that piece, thanks!

**Comment by lucymcnatt** [comment]
Liz covered most of the questions I would've had, my only suggestion is for the dashboard, for these 2 time-series dashboards:
image
Could you split them by up/down as the gauge above, or by app? Its a bit hard to read in its current form, otherwise looks great!

**Comment by rtitle** [comment]
Good suggestion @lucymcnatt  - will do. (That's doable in grafana outside of the Leo PR.) Thanks!

**Review by lucymcnatt — APPROVED** [review]

**Comment by rtitle** [comment]
Same RStudio swatomation test failed, but that seems pretty consistent across all test runs.

More interestingly I'm seeing metrics populate for the swat runs from this PR. :-D 

https://grafana.dsp-devops.broadinstitute.org/d/mmgNHw04z/leonardo-apps?orgId=1&from=now-1h&to=now&var-cluster=terra-qa-bees&var-namespace=terra-jenkins-swat-leo-6829&var-uiClient=All&var-cloud=GCP&var-mrg=All

These seem about right, and interesting insight into what the tests are doing
